File: CardMasterSkin/Elements/SECurvedRectangle.vb
Imports System.Drawing.Drawing2D
Imports CardMasterCard.Card
Imports CardMasterSkin.GraphicsElement

Namespace Skins

    Public Class SECurvedRectangle
        Inherits SkinElement

        Private m_curveSize As Integer

        Public Sub New(skin As Skin, width As Integer, height As Integer, curveSize As Integer)
            Me.New(skin, 0, 0, width, height, curveSize)
        End Sub

        Public Sub New(skin As Skin, x As Integer, y As Integer, width As Integer, height As Integer, curveSize As Integer)
            MyBase.New(skin, x, y, width, height)

            m_curveSize = curveSize

        End Sub

        Public Function GetCurveSize() As Integer
            Return m_curveSize
        End Function

        Public Sub SetCurveSize(curveSize As Integer)
            m_curveSize = curveSize
        End Sub

        Protected Overrides Function GetGraphicElements(card As Card) As List(Of GraphicElement)

            Dim w As Integer = Width
            Dim h As Integer = Height
            Dim c As Integer = GetCurveSize()
            Dim d As Integer = 2 * c

            Dim graphicElementsList As New List(Of GraphicElement)()
            Dim path As New GraphicsPath()

            path.StartFigure()

            ' Côté gauche
            path.AddArc(X, Y, d, h, 90, 180)
            ' Côté haut
            path.AddLine(X + c, Y, X + w - c, Y)
            ' Côté droit
            path.AddArc(X + w - d, Y, d, h, 270, 180)

            ' Côté bas
            path.CloseFigure()

            graphicElementsList.Add(New PathElement(path, GetBackground))

            Return graphicElementsList

        End Function

    End Class

End Namespace

File: CardMasterImageBuilder/Skins/SkinElements/SECurvedRectangle.cs
using CardMasterCard.Card;
using CardMasterImageBuilder.GraphicsElements;
using CardMasterImageBuilder.Skins;
using System.Collections.Generic;
using System.Drawing.Drawing2D;

namespace CardMasterImageBuilder.SkinElements
{
    public class SECurvedRectangle : SkinElement
    {
        private int CurveSize { get; set; } = 0;

        public SECurvedRectangle(Skin skin, int x, int y, int width, int height, string comments, int curveSize) : base(skin, x, y, width, height, comments)
        {
            this.CurveSize = curveSize;
        }

        protected override List<GraphicElement> GetGraphicElements(JsonCard card)
        {
            var graphicElementsList = new List<GraphicElement>();
            var path = new GraphicsPath();

            int w = Width;
            int h = Height;
            int c = this.CurveSize;
            int d = 2 * c;

            // Début de la forme
            path.StartFigure();

            // Côté gauche
            path.AddArc(X, Y, d, h, 90, 180);
            // Côté haut
            path.AddLine(X + c, Y, X + w - c, Y);
            // Côté droit
            path.AddArc(X + w - d, Y, d, h, 270, 180);
            // Côté bas
            // Est généré par la fermeture de la forme

            // Fermeture de la forme
            path.CloseFigure();

            graphicElementsList.Add(new PathElement(path, GetBackground()));

            path = null;

            return graphicElementsList;

        }

    }
}
